### PR TITLE
8313396: Portable implementation of FORBID_C_FUNCTION and ALLOW_C_FUNCTION

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change to how HotSpot prevents the use of certain C library
functions (e.g. poisons references to those functions), while permitting a
subset to be used in restricted circumstances.  Reasons for poisoning a
function include it being considered obsolete, or a security concern, or there
is a HotSpot function (typically in the os:: namespace) providing similar
functionality that should be used instead.

The old mechanism, based on -Wattribute-warning and the associated attribute,
only worked for gcc.  (Clang's implementation differs in an important way from
gcc, which is the subject of a clang bug that has been open for years.  MSVC
doesn't provide a similar mechanism.)  It also had problems with LTO, due to a
gcc bug.

The new mechanism is based on deprecation warnings, using [[deprecated]]
attributes. We redeclare or forward declare the functions we want to prevent
use of as being deprecated.  This relies on deprecation warnings being
enabled, which they already are in our build configuration.  All of our
supported compilers support the [[deprecated]] attribute.

Another benefit of using deprecation warnings rather than warning attributes
is the time when the check is performed.  Warning attributes are checked only
if the function is referenced after all optimizations have been performed.
Deprecation is checked during initial semantic analysis.  That's better for
our purposes here.  (This is also part of why gcc LTO has problems with the
old mechanism, but not the new.)

Adding these redeclarations or forward declarations isn't as simple as
expected, due to differences between the various compilers.  We hide the
differences behind a set of macros, FORBID_C_FUNCTION and related macros.  See
the compiler-specific parts of those macros for details.

In some situations we need to allow references to these poisoned functions.

One common case is where our poisoning is visible to some 3rd party code we
don't want to modify.  This is typically 3rd party headers included in HotSpot
code, such as from Google Test or the C++ Standard Library.  For these the
BEGIN/END_ALLOW_FORBIDDEN_FUNCTIONS pair of macros are used demark the context
where such references are permitted.

Some of the poisoned functions are needed to implement associated HotSpot os::
functions, or in other similarly restricted contexts.  For these, a wrapper
function is provided that calls the poisoned function with the warning
suppressed. These wrappers are defined in the permit_forbidden_functions
namespace, and called using the qualified name. This makes the use of these
functions stand out, suggesting they need careful scrutiny in code reviews and
the like.  There are several benefits to this approach vs the old
ALLOW_C_FUNCTION macro.  We can centralize the set of such functions.  The
syntax for use is simpler (there were syntactic bugs with the old mechanism
that weren't always noticed for a while). The permitted reference is explicit;
there can't be an ALLOW_C_FUNCTION use that isn't actually needed.

Testing:
mach5 tier1-3, which includes various build variants such as slowdebug.
GHA sanity tests
Manual testing for warnings for direct calls to poisoned functions with all 3
compilers, and that the error messages look sane and helpful.

gcc:
```
<file/containing/reference>: In function 'void test_exit(int)':
<file/containing/reference>:<line>:<column>: error: 'void exit(int)' is deprecated: use os::exit [-Werror=deprecated-declarations]
   32 | void test_exit(int status) { return exit(status); }
      |                                     ~~~~^~~~~~~~
```
... and more stuff about the declaration ... 

clang:
```
<file/containing/reference>:<line>:<column>: error: 'exit' is deprecated: use os::exit [-Werror,-Wdeprecated-declarations]
void test_exit(int status) { return exit(status); }
                                    ^
```
... and more stuff about the declaration ... 

Visual Studio:
```
<file/containing/reference>(<line>): warning C4996: 'exit': use os::exit
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8313396](https://bugs.openjdk.org/browse/JDK-8313396): Portable implementation of FORBID_C_FUNCTION and ALLOW_C_FUNCTION (**Enhancement** - P4)

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) Review applies to [b774f14c](https://git.openjdk.org/jdk/pull/22890/files/b774f14c861364d6b76295ebe658555ac4ca3233)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Contributors
 * Martin Doerr `<mdoerr@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22890/head:pull/22890` \
`$ git checkout pull/22890`

Update a local copy of the PR: \
`$ git checkout pull/22890` \
`$ git pull https://git.openjdk.org/jdk.git pull/22890/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22890`

View PR using the GUI difftool: \
`$ git pr show -t 22890`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22890.diff">https://git.openjdk.org/jdk/pull/22890.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22890#issuecomment-2564649321)
</details>

### CODE REVIEW

File: src/hotspot/os/windows/permitForbiddenFunctions_windows.hpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// See forbiddenFunctions.hpp for details.
+
+namespace permit_forbidden_functions {
+BEGIN_ALLOW_FORBIDDEN_FUNCTIONS
+
+// Used by the Windows implementation of os::realpath.
+inline char* _fullpath(char* absPath, const char* relPath, size_t maxLength) {
+  return ::_fullpath(absPath, relPath, maxLength);
+}
+
```
Comment (TheShermanTanker) [inline src/hotspot/os/windows/permitForbiddenFunctions_windows.hpp:41]: Could we forbid \_snprintf too?

Comment (TheShermanTanker) [inline src/hotspot/os/windows/permitForbiddenFunctions_windows.hpp:41]: Could we forbid the non Standard \_snprintf too?

File: src/hotspot/share/utilities/permitForbiddenFunctions.hpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+// Wrapper functions are provided for such forbidden functions.  These
+// wrappers are defined in a context where the forbidding warnings are
+// suppressed.  They are defined in a special namespace, to highlight uses as
+// unusual and requiring increased scrutiny.
+//
+// Note that there are several seemingly plausible shorter alternatives to
+// these written-out wrapper functions.  All that have been tried don't work
+// for one reason or another.
+
+namespace permit_forbidden_functions {
```
Comment (TheShermanTanker) [inline src/hotspot/share/utilities/permitForbiddenFunctions.hpp:48]: permit\_forbidden\_functions sounds like a bit of a mouthful, maybe permit or permitted could work better in this case? If a call to free is to be permitted for example, permit::free(ptr) or permitted::free(ptr) sounds like "permit free", which conveys the intent pretty well, at least in my opinion

File: src/hotspot/share/utilities/compilerWarnings_gcc.hpp [hunk 3]
```diff
@@ -70,32 +70,19 @@
...
+// This variant of FORBID_C_FUNCTION overrides the default variant.  Clang
+// seems to make a distinction between [[noreturn]] and the old-style noreturn
+// attribute.  For example, if <stdlib.h> has already been included, using
+// [[noreturn]] when forbidding exit(int) gives "error: 'noreturn' attribute
+// does not appear on the first declaration", with the previous declaration
+// reported as being "void exit(int) __dead2;". __dead2 is an old-style
+// noreturn attribute.
+#ifdef __clang__
+#define FORBID_NORETURN_C_FUNCTION(Signature, Alternative) \
+  FORBID_C_FUNCTION(__attribute__((__noreturn__)) Signature, Alternative)
```
Comment (TheShermanTanker) [inline src/hotspot/share/utilities/compilerWarnings_gcc.hpp:85]: clang isn't the only one that makes a distinction between [[gnu::noreturn]] and [[noreturn]], gcc does too. It's a little strange that gcc doesn't face this same issue

**Comment by TheShermanTanker** [comment]
Overall nice change, my only complaint is that to know which FORBID macro to use (IMPORTED, NORETURN, etc), you have to delve into implementation details to select the right one. That quickly becomes confusing. But unfortunately all the other approaches thus far have failed, so there aren't really any alternatives that are any better than this approach

File: src/hotspot/share/utilities/compilerWarnings_gcc.hpp (hunk 3 above)
Comment (kimbarrett) [reply src/hotspot/share/utilities/compilerWarnings_gcc.hpp:85]: I don't know what `[[gnu::noreturn]]` has to do with this?  And in what way does gcc treat them differently?

File: src/hotspot/share/utilities/permitForbiddenFunctions.hpp (hunk 2 above)
Comment (kimbarrett) [reply src/hotspot/share/utilities/permitForbiddenFunctions.hpp:48]: It is intentionally a mouthful.  It's not intended to be used except in pretty rare situations, and it should stand
out as unusual and needing additional scrutiny by readers.

File: src/hotspot/os/windows/permitForbiddenFunctions_windows.hpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/os/windows/permitForbiddenFunctions_windows.hpp:41]: Sure, I'll add that.  We should also technically be forbidding snprintf in favor of os::snprintf, but the difference
is pretty minuscule and there are a significant number of occurrences (125 today).

**Comment by kimbarrett** [comment]
> Overall nice change, my only complaint is that to know which FORBID macro to use (IMPORTED, NORETURN, etc), you have to delve into implementation details to select the right one. That quickly becomes confusing. But unfortunately all the other approaches thus far have failed, so there aren't really any alternatives that are any better than this approach

Unfortunately, I couldn't find any way to determine whether IMPORTED or not
other than trial and error.  One would think this might be documented
somewhere, but I didn't find anything.

Whether to use a NORETURN variant seems pretty straightforward.  It would be
better if we could just say (for example)
```
FORBID_C_FUNCTION([[noreturn]] void exit(int), "use os::exit")
```
But that doesn't work both because of the clang weirdness, and because all
attributes must preceed the Windows dllimport spec.

We'll have to come up with something new if we ever want to forbid a function
that has other attributes that we need to mention. I'm hoping that won't
happen, but I haven't (yet) gone through any secure coding guidelines looking
for additional functions to forbid.

One option would be to have distinct arguments for the signature, like we do
for JNI/JVM entry points, e.g.
```
FORBID_C_FUNCTION(char*, strerror, (int))
```
with attributes at the end or something like that.

File: src/hotspot/share/utilities/compilerWarnings_gcc.hpp (hunk 3 above)
Comment (TheShermanTanker) [reply src/hotspot/share/utilities/compilerWarnings_gcc.hpp:85]: [[gnu::noreturn]] is equal to \_\_attribute\_\_((noreturn)), I just referred to it as its scoped attribute form to make it clear that I wasn't talking about the Standard noreturn. There is funnily enough no documentation on how gcc treats them differently, I only learnt that it does while reading through gcc's source code one day

```cpp
/* We used to treat C++11 noreturn attribute as equivalent to GNU's,
	 but no longer: we have to be able to tell [[noreturn]] and
	 __attribute__((noreturn)) apart.
	 Similarly for C++14 deprecated attribute, we need to emit extra
	 diagnostics for [[deprecated]] compared to [[gnu::deprecated]].  */
      /* C++17 fallthrough attribute is equivalent to GNU's.  */
```

Comment (kimbarrett) [reply src/hotspot/share/utilities/compilerWarnings_gcc.hpp:85]: gcc does treat [[noreturn]] differently from [[gnu::noreturn]] /
attribute((noreturn)) when there is a preceding declaration that doesn't have
either kind of attribute.  Compile this:
```
void frob(int);
//__attribute__((__noreturn__)) void frob(int);
//[[gnu::noreturn]] void frob(int);
[[noreturn]] void frob(int);
```
and you'll get an error:
```
error: function 'void frob(int)' declared '[[noreturn]]' but its first declaration was not
```
But uncomment either of the gnu-specific attributes and the error goes away.

C++14 7.6.3 "Noreturn attribute" says "The first declaration of a function
shall specify the noreturn attribute if any declaration of that function
specifies the noreturn attribute."

So without the declarations with gcc-specific attributes it errors, and that's
fine. The gcc-specific attributes never complained in that situation, and
continue to not complain, for backward compatibility. But other than that
error checking difference, the gcc-specific attributes seem to be treated as
equivalent to the standard attribute.

Comment (kimbarrett) [reply src/hotspot/share/utilities/compilerWarnings_gcc.hpp:85]: The difference between [[deprecated]] and [[gnu::deprecated]] is discussed here:
https://gcc.gnu.org/git/?p=gcc.git;a=commit;f=gcc/cp/parser.cc;h=cd647514a539943ade6461efbf056a7c3f4305c6
TL;DR - [[gnu::deprecated]] is allowed in more places than [[deprecated]].

File: src/hotspot/os/linux/mallocInfoDcmd.cpp [hunk 4]
```diff
@@ -36,15 +37,15 @@ void MallocInfoDcmd::execute(DCmdSource source, TRAPS) {
 #ifdef __GLIBC__
   char* buf;
   size_t size;
-  ALLOW_C_FUNCTION(::open_memstream, FILE* stream = ::open_memstream(&buf, &size);)
+  FILE* stream = ::open_memstream(&buf, &size);
```
Comment (kimbarrett) [inline src/hotspot/os/linux/mallocInfoDcmd.cpp:40]: Note that open_memstream was never forbidden, so the ALLOW_C_FUNCTION was not needed.

File: src/hotspot/os/linux/mallocInfoDcmd.cpp [hunk 5]
```diff
@@ -36,15 +37,15 @@ void MallocInfoDcmd::execute(DCmdSource source, TRAPS) {
...
   if (stream == nullptr) {
     _output->print_cr("Error: Could not call malloc_info(3)");
     return;
   }
 
   int err = os::Linux::malloc_info(stream);
   if (err == 0) {
-    ALLOW_C_FUNCTION(::fflush, fflush(stream);)
+    fflush(stream);
```
Comment (kimbarrett) [inline src/hotspot/os/linux/mallocInfoDcmd.cpp:48]: Note that fflush was never forbidden, so the ALLOW_C_FUNCTION was not needed.

File: src/hotspot/os/linux/mallocInfoDcmd.cpp [hunk 6]
```diff
@@ -54,8 +55,8 @@ void MallocInfoDcmd::execute(DCmdSource source, TRAPS) {
   } else {
     ShouldNotReachHere();
   }
-  ALLOW_C_FUNCTION(::fclose, ::fclose(stream);)
-  ALLOW_C_FUNCTION(::free, ::free(buf);)
+  ::fclose(stream);
```
Comment (kimbarrett) [inline src/hotspot/os/linux/mallocInfoDcmd.cpp:58]: Note that fclose was never forbidden, so the ALLOW_C_FUNCTION was not needed.

**Comment by TheRealMDoerr** [comment]
Unfortunately, this doesn't compile on AIX:
```
globalDefinitions_gcc.hpp:42:
In file included from /opt/IBM/openxlC/17.1.1/bin/../include/c++/v1/stdlib.h:93:
/usr/include/stdlib.h:304:18: error: 'noreturn' attribute does not appear on the first declaration
        extern _NOTHROW(_NORETURN(void, exit), (int));
                        ^
/usr/include/comp_macros.h:30:35: note: expanded from macro '_NORETURN'
#define _NORETURN(_T, _F) _T _F [[noreturn]]
                                  ^
   ... (rest of output omitted)
```

**Comment by kimbarrett** [comment]
> Unfortunately, this doesn't compile on AIX: [...]

Ugh!  The clang workaround  for noreturn handling is going to need to be more extensive.

**Comment by jdksjolen** [comment]
Hi, 

I really like the syntactic change of this feature, and it's very nice that we get to have working auto-complete (makes it easier to find out that a specific function isn't forbidden). The syntactic change isn't shown in the PR description, isn't that useful to add?

I have a bike shedding request: Could we skip the S at the end and have it be `permit_forbidden_function::free(my_ptr);`? We only allow one forbidden function in that call, so this reads neater.

**Comment by TheShermanTanker** [comment]
Hi Johan, which syntactic change are you referring to?

**Comment by jdksjolen** [comment]
> Hi Johan, which syntactic change are you referring to?

Oh, just that we don't have to repeat the name of the function we want to use. No more macro, just a namespace and a function!

File: src/hotspot/share/utilities/compilerWarnings.hpp [hunk 7]
```diff
@@ -86,25 +86,58 @@
 #define PRAGMA_ZERO_AS_NULL_POINTER_CONSTANT_IGNORED
 #endif
 
+//////////////////////////////////////////////////////////////////////////////
```
Comment (jdksjolen) [inline src/hotspot/share/utilities/compilerWarnings.hpp:89]: This doesn't adhere to the style of the file AFAICS and is unnecessary, can we remove it?

**Review by jdksjolen — COMMENTED** [review]
I've read through it all, and it seems like good code to me. I like the change in syntax a lot. I've one nit regarding comment style.

File: src/hotspot/share/utilities/forbiddenFunctions.hpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+#ifndef SHARE_UTILITIES_FORBIDDENFUNCTIONS_HPP
+#define SHARE_UTILITIES_FORBIDDENFUNCTIONS_HPP
+
+#include "utilities/compilerWarnings.hpp"
+#include "utilities/macros.hpp"
+#include <stdarg.h> // for va_list
+#include <stddef.h> // for size_t
+
+#include OS_HEADER(forbiddenFunctions)
```
Comment (coleenp) [inline src/hotspot/share/utilities/forbiddenFunctions.hpp:33]: I thought there was an OS_VARIANT_HEADER to dispatch directly to the posix variant, but I guess it doesn't exist.  Using the semaphore example, this could save your dispatch header files, at the cost of an #ifdef.  Not sure which is worse:

```
#if defined(LINUX) || defined(AIX)
# include "semaphore_posix.hpp"
#else
# include OS_HEADER(semaphore)
#endif
```

**Review by coleenp — COMMENTED** [review]
This is very nice, now that I've navigated to all the pieces and see how it works.  I would rather see #ifdef include the posix file rather than the dispatch header files which would help finding all the bits again.

File: src/hotspot/share/utilities/forbiddenFunctions.hpp (hunk 8 above)
Comment (coleenp) [reply src/hotspot/share/utilities/forbiddenFunctions.hpp:33]: park.hpp has the same:

```
#if defined(LINUX) || defined(AIX) || defined(BSD)
# include "park_posix.hpp"
#else
# include OS_HEADER(park)
#endif
```

**Comment by kimbarrett** [comment]
> Unfortunately, this doesn't compile on AIX:
> 
> ```
> globalDefinitions_gcc.hpp:42:
> In file included from /opt/IBM/openxlC/17.1.1/bin/../include/c++/v1/stdlib.h:93:
> /usr/include/stdlib.h:304:18: error: 'noreturn' attribute does not appear on the first declaration
>         extern _NOTHROW(_NORETURN(void, exit), (int));
>                         ^
> /usr/include/comp_macros.h:30:35: note: expanded from macro '_NORETURN'
> #define _NORETURN(_T, _F) _T _F [[noreturn]]
>                                   ^
>    ... (rest of output omitted)
> ```

Thanks for testing that port.

I restructured the implementation of FORBID_C_FUNCTION, and added more commentary about the clang issue.
I think that didn't change the resulting expansion, but I think made it easier to describe how I tried to address this
problem.  The actual "solution" (I hope) is to ensure that <stdlib.h> is included before the forbidding declarations
for the noreturn functions.  Please try an aix build again.

**Comment by kimbarrett** [comment]
> I really like the syntactic change of this feature, and it's very nice that we get to have working auto-complete (makes it easier to find out that a specific function isn't forbidden). The syntactic change isn't shown in the PR description, isn't that useful to add?

Isn't the last paragraph of the PR description (beginning with "Some of the poisoned functions...") what your are
looking for?  Admittedly I didn't mention working with auto-complete as one of the benefits.

> I have a bike shedding request: Could we skip the S at the end and have it be `permit_forbidden_function::free(my_ptr);`? We only allow one forbidden function in that call, so this reads neater.

I named it as a collection of forbidden functions.  But I agree it reads better without the plural, so I've removed the S.

File: src/hotspot/share/utilities/forbiddenFunctions.hpp (hunk 8 above)
Comment (kimbarrett) [reply src/hotspot/share/utilities/forbiddenFunctions.hpp:34]: If I was going to go that route, I'd be more inclined toward
```
#if !define(_WINDOWS)
#include "forbiddenFunctions_windows.hpp"
#else
#include "forbiddenFunctions_posix.hpp"
#endif
```
rather than list all (or maybe just a subset) of the posix-like ports.  My
inclination is to leave it as-is with OS_HEADERS, since I think that's the
"intended" idiom.

**Comment by TheRealMDoerr** [comment]
> > Unfortunately, this doesn't compile on AIX:
> > ```
> > globalDefinitions_gcc.hpp:42:
> > In file included from /opt/IBM/openxlC/17.1.1/bin/../include/c++/v1/stdlib.h:93:
> > /usr/include/stdlib.h:304:18: error: 'noreturn' attribute does not appear on the first declaration
> >         extern _NOTHROW(_NORETURN(void, exit), (int));
> >                         ^
> > /usr/include/comp_macros.h:30:35: note: expanded from macro '_NORETURN'
> > #define _NORETURN(_T, _F) _T _F [[noreturn]]
> >                                   ^
> >    ... (rest of output omitted)
> > ```
> 
> Thanks for testing that port.
> 
> I restructured the implementation of FORBID_C_FUNCTION, and added more commentary about the clang issue. I think that didn't change the resulting expansion, but I think made it easier to describe how I tried to address this problem. The actual "solution" (I hope) is to ensure that <stdlib.h> is included before the forbidding declarations for the noreturn functions. Please try an aix build again.

Thanks! This has solved one of the problems, but not all. The next one is:
```
globalDefinitions_gcc.hpp:55:
In file included from /usr/include/fcntl.h:242:
/usr/include/unistd.h:181:8: error: 'noreturn' attribute does not appear on the first declaration
extern _NORETURN(void, _exit)(int);
       ^
/usr/include/comp_macros.h:30:35: note: expanded from macro '_NORETURN'
#define _NORETURN(_T, _F) _T _F [[noreturn]]
```

**Comment by kimbarrett** [comment]
> Thanks! This has solved one of the problems, but not all. The next one is:
> 
> ```
> globalDefinitions_gcc.hpp:55:
> In file included from /usr/include/fcntl.h:242:
> /usr/include/unistd.h:181:8: error: 'noreturn' attribute does not appear on the first declaration
> extern _NORETURN(void, _exit)(int);
>        ^

Drat.  I forgot that posix _exit comes from unistd.h.  Back to work...

File: src/hotspot/share/utilities/forbiddenFunctions.hpp (hunk 8 above)
Comment (dholmes-ora) [reply src/hotspot/share/utilities/forbiddenFunctions.hpp:34]: Overall I like this change. I appreciate the effort that has been put in to try and find an elegant solution to this problem.

but having OS specific files created just to include the posix version runs counter to why we have the posix variants in the first place IMO. Please select one of the above approaches so that the new aix/bsd/linux specific files can be removed in favour of the posix one. Thanks.

**Comment by kimbarrett** [comment]
> > Thanks! This has solved one of the problems, but not all. The next one is:
> > ```
> > globalDefinitions_gcc.hpp:55:
> > In file included from /usr/include/fcntl.h:242:
> > /usr/include/unistd.h:181:8: error: 'noreturn' attribute does not appear on the first declaration
> > extern _NORETURN(void, _exit)(int);
> >        ^
> > ```
> 
> Drat. I forgot that posix _exit comes from unistd.h. Back to work...

Hopefully that's fixed now too, and there won't be any more.
I also added _Exit to the forbidden set.

I looked into the more robust / less kludgy approaches, but they are more work than I want to do as part of this PR.
Maybe in a followup.  And I wonder if someone might file a bug about the clang weirdness?

This might be related, but seems kind of different:
https://github.com/llvm/llvm-project/issues/113511

Test case:
```
__attribute__((__noreturn__)) void frob(int);
[[noreturn]] void frob(int);
```
=>
```
'noreturn' attribute does not appear on the first declaration
```
gcc compiles this without error.

**Comment by kimbarrett** [comment]
> Thanks! That solved the issue. Now, we need to adapt all usages in the low level AIX code and support `strdup`:

Not surprising that there are a few of these, since the aix-specific code
wasn't previously under the forbidden-function regime.

I've applied most of the provided patches. (And thanks for providing them.)
The exception is the patches related to strdup. There's only one use, in
loadlib_aix.cpp. Rather than the patch to add a permit wrapper for strdup
(affecting all other code too) and use that here, I'd rather just suppress the
warning explicitly here.

I filed an enhancement to remove the use of ::strdup.
https://bugs.openjdk.org/browse/JDK-8347157

I also filed this after looking for strdup uses in aix code.
https://bugs.openjdk.org/browse/JDK-8347143
It's about a strange use of os::strdup, so doesn't affect this PR.

> It may be possible to use some `os::` versions, but that would need a careful review. I think it's safer to keep it this way for now. @JoKern65: You may want to take a look, too.

I'll leave it to you folks to file any additional issues related to using
permit bypass functions instead of the relevant os:: functions.

File: src/hotspot/share/utilities/forbiddenFunctions.hpp (hunk 8 above)
Comment (kimbarrett) [reply src/hotspot/share/utilities/forbiddenFunctions.hpp:34]: I disagree. It seems to me that breaking the abstraction like that is just asking for trouble.

**Comment by TheRealMDoerr** [comment]
Thanks! The AIX build was successful with the recent changes.
Regarding [JDK-8347157](https://bugs.openjdk.org/browse/JDK-8347157), I don't think we should use `os::strdup` in `StringList` because the class is designed to use raw malloc. I guess @JoKern65 can take a look at the 2 RFEs.

**Review by jdksjolen — APPROVED** [review]
If this passes testing and David's comments are fixed, then I'm very happy with this PR as it is. Thank you for this.

File: src/hotspot/share/utilities/forbiddenFunctions.hpp (hunk 8 above)
Comment (coleenp) [reply src/hotspot/share/utilities/forbiddenFunctions.hpp:34]: Yes, I think it's fine to say !WINDOWS instead of listing all the posix ports.  We've been avoiding dispatch files and once it reaches a threshold of too many #ifdef !WINDOWS #include posix.hpp one, then we could add another macro like the OS_CPU one.

Also the copyright script added 2025 for these because they started with 2024 so it's sort of a bug in the script but not really solvable because it doesn't know that you didn't check this in in 2024.

Comment (dholmes-ora) [reply src/hotspot/share/utilities/forbiddenFunctions.hpp:34]: > It seems to me that breaking the abstraction like that is just asking for trouble.

Sorry but what "abstraction" are you referring to?

Comment (kimbarrett) [reply src/hotspot/share/utilities/forbiddenFunctions.hpp:34]: @dholmes-ora - OS_HEADER and CPU_HEADER instead of explicit knowledge of what platforms exist and
which ones share some code in a "_posix" file and which ones have some unshared code.  IMO the includer
shouldn't need to know that kind of implementation detail.

But oh well, all y'all seem to really hate the simple forwarding files and would prefer to skip that in favor of
hard-coding the file organization.  In the interest of making progress, I'll do that.

Comment (kimbarrett) [reply src/hotspot/share/utilities/forbiddenFunctions.hpp:34]: @coleenp The copyrights on the new files being 2024-2025 is intentional. They were first published (in this PR)
in 2024, so I think are supposed to have that starting year.

**Comment by kimbarrett** [comment]
> I filed an enhancement to remove the use of ::strdup. https://bugs.openjdk.org/browse/JDK-8347157

Per discussion in that bug, it's been closed as not an issue, and I'm changing the code here to add and use the
permit wrapper for strdup.

File: src/hotspot/os/posix/forbiddenFunctions_posix.hpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef OS_POSIX_FORBIDDENFUNCTIONS_POSIX_HPP
+#define OS_POSIX_FORBIDDENFUNCTIONS_POSIX_HPP
+
+#include "utilities/compilerWarnings.hpp"
+#include <stddef.h> // for size_t
```
Comment (stefank) [inline src/hotspot/os/posix/forbiddenFunctions_posix.hpp:29]:
```suggestion
#include "utilities/compilerWarnings.hpp"

#include <stddef.h> // for size_t
```

File: src/hotspot/os/windows/forbiddenFunctions_windows.hpp [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef OS_WINDOWS_FORBIDDENFUNCTIONS_WINDOWS_HPP
+#define OS_WINDOWS_FORBIDDENFUNCTIONS_WINDOWS_HPP
+
+#include "utilities/compilerWarnings.hpp"
+#include <stddef.h> // for size_t
```
Comment (stefank) [inline src/hotspot/os/windows/forbiddenFunctions_windows.hpp:29]:
```suggestion
#include "utilities/compilerWarnings.hpp"

#include <stddef.h> // for size_t
```

File: src/hotspot/share/utilities/forbiddenFunctions.hpp [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_FORBIDDENFUNCTIONS_HPP
+#define SHARE_UTILITIES_FORBIDDENFUNCTIONS_HPP
+
+#include "utilities/compilerWarnings.hpp"
+#include "utilities/macros.hpp"
+#include <stdarg.h> // for va_list
```
Comment (stefank) [inline src/hotspot/share/utilities/forbiddenFunctions.hpp:30]:
```suggestion
#include "utilities/macros.hpp"

#include <stdarg.h> // for va_list
```

File: src/hotspot/share/utilities/permitForbiddenFunctions.hpp [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+inline void* calloc(size_t nmemb, size_t size) { return ::calloc(nmemb, size); }
+inline void* realloc(void* ptr, size_t size) { return ::realloc(ptr, size); }
+
+inline char* strdup(const char* s) { return ::strdup(s); }
+
+END_ALLOW_FORBIDDEN_FUNCTIONS
+} // namespace permit_forbidden_function
+
+#endif // SHARE_UTILITIES_PERMITFORBIDDENFUNCTIONS_HPP
+
```
Comment (stefank) [inline src/hotspot/share/utilities/permitForbiddenFunctions.hpp:70]:
```suggestion
#endif // SHARE_UTILITIES_PERMITFORBIDDENFUNCTIONS_HPP
```

File: test/hotspot/gtest/unittest.hpp [hunk 13]
```diff
@@ -24,6 +24,7 @@
 #ifndef UNITTEST_HPP
 #define UNITTEST_HPP
 
+#include "utilities/globalDefinitions.hpp"
```
Comment (stefank) [inline test/hotspot/gtest/unittest.hpp:27]:
```suggestion
#include "utilities/globalDefinitions.hpp"

```

**Comment by kimbarrett** [comment]
@stefank Someday that long ago long discussion about include ordering and formatting needs to get cleaned up and
added to the style guide.

File: src/hotspot/os/bsd/permitForbiddenFunctions_bsd.hpp [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+ * questions.
+ *
+ */
+
+#ifndef OS_BSD_PERMITFORBIDDENFUNCTIONS_BSD_HPP
+#define OS_BSD_PERMITFORBIDDENFUNCTIONS_BSD_HPP
+
+#include "permitForbiddenFunctions_posix.hpp"
+
+#endif // OS_BSD_PERMITFORBIDDENFUNCTIONS_BSD_HPP
```
Comment (coleenp) [inline src/hotspot/os/bsd/permitForbiddenFunctions_bsd.hpp:30]: I thought you were eliminating these in favor of #ifdef where included?

Comment (kimbarrett) [reply src/hotspot/os/bsd/permitForbiddenFunctions_bsd.hpp:30]: Removed the forbid files, forgot the permit files.  Sigh!  I'll push a new commit after running tests.

File: src/hotspot/share/utilities/forbiddenFunctions.hpp (hunk 8 above)
Comment (dholmes-ora) [reply src/hotspot/share/utilities/forbiddenFunctions.hpp]: @kimbarrett I understand. I also don't like the includer needing to know about this, and maybe it is time to add OS_FAMILY_HEADER to deal with it. But in the absence of the new macro I prefer this break of abstraction to the creation of a bunch of tiny forwarding files.

Comment (coleenp) [reply src/hotspot/share/utilities/forbiddenFunctions.hpp]: I prefer the opposite.  If you break these up into a bunch of forwarding files, do all the rest of the VM the same way.

Comment (dholmes-ora) [reply src/hotspot/share/utilities/forbiddenFunctions.hpp]: @coleenp I thought you and I were on the same page not liking the creation of all these little files that just forward to the posix one. ???

**Review by dholmes-ora — APPROVED** [review]
Nothing further from me.

Thanks

**Comment by kimbarrett** [comment]
> This looks good for me too. I think you should allow GHA to run for this change.

I ran GHA tests on at least one early version, but I'll do another run before integrating.

**Comment by kimbarrett** [comment]
Thanks for reviews.

**Comment by TheShermanTanker** [comment]
Curious question: Did we ever find out why the attribute warning didn't fire off at the sites of forbidden methods being called inside gtest?
